src/ocl/rnnocl.cpp

```diff
@@ -3791,131 +4392,760 @@ void RNNDescriptor::RNNBackwardData(Handle& handle,
+
+        miopen::GemmDescriptor gemm_desc = GemmDescriptor{false,
+                                                          false,
+                                                          false,
+                                                          batch_size,
+                                                          hidden_size,
+                                                          hidden_size,
+                                                          RBuff.gemm_write_size(),
+                                                          hidden_size,
+                                                          hidden_size,
+                                                          1, // batch count
+                                                          0, // Stride A
+                                                          0, // Stride B
+                                                          0, // Stride C
+                                                          1, // alpha
+                                                          1, // beta
+                                                          rnn_data_type,
+                                                          false};
+
+        int output_batch = time == 0 ? 0 : batches.at(prev_time);
+
+        int input_batch =
+            time == 0 ? bacc_per_time[cur_time] : bacc_per_time[prev_time] - batch_size;
+
+        miopenStatus_t gemm_status =
+            CallGemm(handle,
+                     gemm_desc,
+                     workSpace,
+                     RBuff.gemm_write_offset(layer, input_batch, direction),
+                     w,
+                     WeiBuf.hidden_weight_offset(layer, direction),
+                     dhx,
+                     get_HxBuff_offset(layer, output_batch, direction),
+                     GemmBackend_t::rocblas);
+
+        if(gemm_status != miopenStatusSuccess)
+        {
+            if(gemm_status == miopenStatusNotImplemented)
             {
-                int batchval, batchvalout;
-                std::tie(batchval, std::ignore)    = miopen::tien<2>(dxDesc[i].GetLengths());
-                std::tie(batchvalout, std::ignore) = miopen::tien<2>(dyDesc[i].GetLengths());
-                if(batchval != batchvalout)
-                {
-                    MIOPEN_THROW(miopenStatusBadParm,
-                                 "Input batch length: " + std::to_string(batchval) +
-                                     ", Output batch length: " + std::to_string(batchvalout));
-                }
-                in_n[i] = batchval;
+                MIOPEN_LOG_E("GEMM not implemented");
+            }
+            else
+            {
+                MIOPEN_LOG_E("GEMM failed");
             }
+        }
+    };
 
-            RNNTensorPaddingConverter::ConvertTensorData(
-                handle, dyDesc[0], in_n, dy, packedDYIn, true);
+    auto propagate_dhx = [*this, seqLen, &propagate_dhx_prev, &dhx](int layer) {
```

Comment:
`***this** `
copy-capture for current class? really?

`&dhx`
reference to ptr?

Reply:
fixed